server/aesdsocket.c

```diff
@@ -12,43 +12,139 @@
 #include <fcntl.h>
 #include <signal.h>
 
+#include <stdbool.h>
+#include <pthread.h>
+#include <sys/signal.h>
+#include <sys/time.h>
+#include <time.h>
+#include "queue.h"
+
 #define MAX_BUF_SIZE         (1024)
 #define PORT                 (9000)
 #define SA struct sockaddr   
 
-#if 0
-int getaddrinfo(const char *restrict node,
-                       const char *restrict service,
-                       const struct addrinfo *restrict hints,
-                       struct addrinfo **restrict res);
+// client connection queue lock
+pthread_mutex_t queue_lock;
+
+// lock to protect pkt file
+pthread_mutex_t file_lock;
+
+//client connection queue
+typedef struct client_queue {
+	int        connfd;           //client connection fd
+	int        task_finished;    //flag to indicate client thread is done
+	pthread_t  threadId;
+	TAILQ_ENTRY(client_queue) ent;
+} client_queue_t;
+
+typedef TAILQ_HEAD(head_s, client_queue) head_t;
+
+//head of the queue
+head_t head;
+
+//thread function to process the client's connection
+void *process_client_connection(void *e);
+
+// Takes a string and puts in in the queue on character at a time.
+static int create_client_thread(int connfd)
+{
+    int rc = -1; //return code
+
+    struct client_queue *e = malloc(sizeof(struct client_queue));
+    if (e == NULL)
+    {
+        syslog(LOG_ERR, "Failed to allocate memory");
+	return -1;
+    }
+
+    e->connfd = connfd;
+    e->task_finished = 0;
+
+    // Create a thread that will function threadFunc()
+    rc = pthread_create(&e->threadId, NULL, &process_client_connection, (void *)e);
+    if (rc != 0) {
+        free(e); // free the allocated memory and return error
+        return -1;
+    }
+
+    //acquire lock
+    if (pthread_mutex_lock(&queue_lock) != 0) {
+        syslog(LOG_ERR, "\r\n Failed to acquire lock"); //assert would be better
+	return -1;
+    }
+    TAILQ_INSERT_TAIL(&head, e, ent);
```

Comment:
@maheshchnt I don't think you need to lock `head` here but if you did you should make it more clear that this lock protects the `head` item.  Another reason to move this from a global variable, to make it clear that it's not accessed outside the thread.